### PR TITLE
Fix transition issues

### PR DESCRIPTION
**Pull request by zsoltk**
Main improvements:

- Added new abstraction on how Actions are reversed (addressing main issue)
- Effect emission moved to responsibility of Actions, now when the Action is reversed, so is the Effect (addressing main issue)
- Removed special case of Add resolution (+NoOpAction), it's now same approach with all other Actions, and can be reversed correctly (something not true to previous fixed Add call where NoOp was being reversed)
- Covered edge cases of pending detach / removal (unfinished transition) + save/restore cycle
- Configuration <C> moved to ConfigurationKey completely, no longer needed in ConfigurationCommand (lighter code)

Should work correctly and without a hiccup in all cases of push / replace and save / restore cycles too.

### CODE REVIEW

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/transition/handler/Defaults.kt [hunk 1]
```diff
@@ -3,6 +3,6 @@ package com.badoo.ribs.core.routing.transition.handler
 import android.view.animation.AccelerateDecelerateInterpolator
 import android.view.animation.Interpolator
 
-internal const val defaultDuration: Long = 300
+internal const val defaultDuration: Long = 3000
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/transition/handler/Defaults.kt]: Reduce back to 300 ?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/OngoingTransition.kt [hunk 2]
```diff
@@ -63,6 +63,7 @@ internal class OngoingTransition<C : Parcelable>(
     fun reverse() {
         transitionPair.exiting?.reverse()
         transitionPair.entering?.reverse()
+        actions = actions.reversed()
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/OngoingTransition.kt]: lol

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/ConfigurationFeatureState.kt [hunk 3]
```diff
@@ -43,20 +47,23 @@ internal data class WorkingState<C : Parcelable>(
...
-                    is ConfigurationContext.Unresolved -> entry
-                    is ConfigurationContext.Resolved -> entry.shrink()
-                }.copy(
-                    activationState = it.value.activationState.sleep()
-                )
-            }.toMap()
+            pool.filter { !pendingRemoval.contains(it.key) }
+                .map { entry ->
+                    val original = entry.value
+                    val detachRespected = when {
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/ConfigurationFeatureState.kt]: detached?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/ConfigurationFeatureActor.kt [hunk 4]
```diff
@@ -165,53 +153,61 @@ internal class ConfigurationFeatureActor<C : Parcelable>(
...
+        defaultElements: Pool<C>,
+        transaction: Transaction<C>? = null
+    ): TransactionExecutionParams<C> {
+        val tempPool: MutablePool<C> = state.pool.toMutablePool()
+
+        return TransactionExecutionParams(
+            emitter = emitter,
+            resolver = { key ->
+                val lookup = tempPool[key]
+                if (lookup is ConfigurationContext.Resolved) lookup
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/ConfigurationFeatureActor.kt]: Extract maybe?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/action/multi/WakeUpAction.kt [hunk 5]
```diff
@@ -21,13 +20,24 @@ internal class WakeUpAction<C : Parcelable> : MultiConfigurationAction<C> {
      */
     override fun execute(
         state: WorkingState<C>,
-        params: ActionExecutionParams<C>
-    ): Map<ConfigurationKey, ConfigurationContext.Resolved<C>> =
-        state.pool.invokeOn(SLEEPING, params) { foundByFilter ->
-            val action = ActivateAction(foundByFilter, params, false)
+        params: TransactionExecutionParams<C>
+    ) {
+        state.pool.invokeOn(SLEEPING, params) { key, foundByFilter ->
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/action/multi/WakeUpAction.kt]: maybe just `key, configuration`? to mention the type of second parameter in its name

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/transition/handler/Defaults.kt (hunk 1 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/transition/handler/Defaults.kt]: yeah

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/OngoingTransition.kt (hunk 2 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/OngoingTransition.kt]: learning from the best :P

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/ConfigurationFeatureActor.kt (hunk 4 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/ConfigurationFeatureActor.kt]: Tried, then reverted for increasing overall complexity how it ended up.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/action/multi/WakeUpAction.kt (hunk 5 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/action/multi/WakeUpAction.kt]: Sure. It's not a `Configuration` though but a `ConfigurationContext` (think Resolved / Unresolved), which itself should have a better name.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/ConfigurationFeatureState.kt (hunk 3 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/ConfigurationFeatureState.kt]: updated

**Pull request merged**